node/src/command/solochain.rs

```diff
@@ -0,0 +1,415 @@
+    jsonrpsee::server::BatchRequestConfig,
+    log::{info, warn},
+    sc_chain_spec::{ChainType, GenericChainSpec, NoExtension},
+    sc_cli::{CliConfiguration, DefaultConfigurationValues, Signals, SubstrateCli},
+    sc_network::config::{NetworkBackendType, NetworkConfiguration, TransportConfig},
+    sc_network_common::role::Role,
+    sc_service::{
+        config::KeystoreConfig, BasePath, BlocksPruning, Configuration, DatabaseSource, TaskManager,
+    },
+    sc_tracing::logging::LoggerBuilder,
+    std::{
+        future::Future,
+        num::NonZeroUsize,
+        path::{Path, PathBuf},
+        time::Duration,
+    },
+    tc_service_container_chain::cli::ContainerChainCli,
+};
+
+/// Alternative to [Configuration] struct used in solochain context.
+pub struct SolochainConfig {
+    pub tokio_handle: tokio::runtime::Handle,
+    pub base_path: BasePath,
+    pub network_node_name: String,
+    pub role: Role,
+    pub relay_chain: String,
+}
+
+/// Alternative to [Runner](sc_cli::Runner) struct used in solochain context.
+pub struct SolochainRunner {
+    config: SolochainConfig,
+    tokio_runtime: tokio::runtime::Runtime,
+    signals: Signals,
+}
+
+impl SolochainRunner {
+    /// Log information about the node itself.
+    ///
+    /// # Example:
+    ///
+    /// ```text
+    /// 2020-06-03 16:14:21 Substrate Node
+    /// 2020-06-03 16:14:21 ✌️  version 2.0.0-rc3-f4940588c-x86_64-linux-gnu
+    /// 2020-06-03 16:14:21 ❤️  by Parity Technologies <admin@parity.io>, 2017-2020
+    /// 2020-06-03 16:14:21 📋 Chain specification: Flaming Fir
+    /// 2020-06-03 16:14:21 🏷  Node name: jolly-rod-7462
+    /// 2020-06-03 16:14:21 👤 Role: FULL
+    /// 2020-06-03 16:14:21 💾 Database: RocksDb at /tmp/c/chains/flamingfir7/db
+    /// 2020-06-03 16:14:21 ⛓  Native runtime: node-251 (substrate-node-1.tx1.au10)
+    /// ```
+    fn print_node_infos(&self) {
+        use chrono::offset::Local;
+        use chrono::Datelike;
+        type C = ContainerChainCli;
+        info!("{}", C::impl_name());
+        info!("✌️  version {}", C::impl_version());
+        info!(
+            "❤️  by {}, {}-{}",
+            C::author(),
+            C::copyright_start_year(),
+            Local::now().year()
+        );
+        // No chain spec
+        //info!("📋 Chain specification: {}", config.chain_spec.name());
+        info!("🏷  Node name: {}", self.config.network_node_name);
+        info!("👤 Role: {}", self.config.role);
+        info!(
+            "💾 Database: {} at {}",
+            // Container chains only support paritydb
+            "ParityDb",
+            // Print base path instead of db path because each container will have its own db in a
+            // different subdirectory.
+            self.config.base_path.path().display(),
+        );
+    }
+
+    /// A helper function that runs a node with tokio and stops if the process receives the signal
+    /// `SIGTERM` or `SIGINT`.
+    pub fn run_node_until_exit<F, E>(
```

Comment:
is this copied from polkadot?

Reply:
Yes, most of the functions in this file are

Reply:
*disregard*